File: java/src/main/java/it/giannibombelli/wsc2026/payment/application/usecases/RefundRequesting.java
package it.giannibombelli.wsc2026.payment.application.usecases;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.common.application.UseCase;
import it.giannibombelli.wsc2026.common.application.events.EventPublisher;
import it.giannibombelli.wsc2026.payment.application.commands.RefundTransaction;
import it.giannibombelli.wsc2026.payment.domain.events.PaymentEvent;
import it.giannibombelli.wsc2026.payment.domain.events.RefundRequested;
import it.giannibombelli.wsc2026.payment.domain.payment.Payment;
import it.giannibombelli.wsc2026.payment.domain.ports.PaymentRepository;

import static java.util.Objects.requireNonNull;

public final class RefundRequesting implements UseCase<RefundTransaction, RefundRequested> {
    private final PaymentRepository paymentRepository;
    private final EventPublisher<PaymentEvent> eventPublisher;

    public RefundRequesting(PaymentRepository paymentRepository, EventPublisher<PaymentEvent> eventPublisher) {
        this.paymentRepository = requireNonNull(paymentRepository);
        this.eventPublisher = requireNonNull(eventPublisher);
    }

    @Override
    public RefundRequested invoke(RefundTransaction cmd) {
        Require.requireArgument(cmd, "command");

        Payment payment = paymentRepository.findById(cmd.aggregateId())
            .orElseThrow(() -> new IllegalArgumentException("payment not found"));

        RefundRequested event = payment.requestRefund(cmd.amount());
        paymentRepository.save(payment);

        eventPublisher.publish(event);
        return event;
    }
}

File: typescript/src/payment/application/usecases/refundRequesting.ts
// Caso d'uso per richiedere un rimborso.

import type { UseCase } from '@/common/application/usecase.js';
import type { EventPublisher } from '@/common/application/events/eventPublisher.js';
import { PaymentEvent } from '../../domain/events/paymentEvent.js';
import { RefundRequested } from '../../domain/events/refundRequested.js';
import { PaymentNotFoundException } from '../../domain/payment/paymentNotFoundException.js';
import type { PaymentRepository } from '../../domain/ports/paymentRepository.js';
import { RefundTransaction } from '../commands/refundTransaction.js';
import { requireArgument } from '@/common/utils/requireArgument.js';
import { requireDependency } from '@/common/utils/requireDependency.js';

export class RefundRequesting implements UseCase<RefundTransaction, RefundRequested> {
  private readonly _paymentRepository: PaymentRepository;
  private readonly _eventPublisher: EventPublisher<PaymentEvent>;

  constructor(paymentRepository: PaymentRepository, eventPublisher: EventPublisher<PaymentEvent>) {
    requireDependency(paymentRepository, "paymentRepository");
    requireDependency(eventPublisher, "eventPublisher");
    this._paymentRepository = paymentRepository;
    this._eventPublisher = eventPublisher;
  }

  invoke(cmd: RefundTransaction): RefundRequested {
    requireArgument(cmd, 'command');

    const payment = this._paymentRepository.findById(cmd.aggregateId);
    if (payment === null) throw new PaymentNotFoundException();

    const event = payment.requestRefund(cmd.amount);
    this._paymentRepository.save(payment);

    this._eventPublisher.publish(event);
    return event;
  }
}
